File: ml-models/MaxVote/python/MaxVote.py
# Imports
import numpy as np
import pandas as pd
import time
import click


@click.group()
def cli():
    pass


@click.command()
def train():
    startTime = time.time()

    with open('latest.model.joblib', 'w') as fp:
        pass
        fp.write("Max Vote Algorithm")

    endTime = time.time()
    timeElapsed = endTime-startTime
    print('Max Vote Ready in %.2fs' % (timeElapsed))
    return timeElapsed


@click.command()
def validate():
    startTime = time.time()

    with open('validate.results.csv', 'w') as fp:
        pass

    endTime = time.time()
    timeElapsed = endTime-startTime
    print('Max Vote Ready in %.2fs' % (timeElapsed))
    return timeElapsed


@click.command()
@click.option('--n_inputs', default=5, required=True)
def vote(n_inputs):
    startTime = time.time()

    # 1. Load dataset
    datasets = []
    for x in range(1, n_inputs+1):
        # Pandas DataFrame
        dataset = loadDataset('predict.' + str(x) + '.csv')
        datasets.append(dataset.iloc[:, 1])

    X = np.column_stack(datasets)

    # 2. Prepare Max Count
    result = []
    for x in range(0, len(X)):
        max = findMajority(X[x], len(X[x]))
        if (max == -1):
            max = X[x][0]
        result.append(max)

    # 3. Store Result
    resultDf = pd.DataFrame(result)
    saveDataset('predict.results.csv', resultDf)

    endTime = time.time()
    timeElapsed = endTime-startTime
    print('Voting Finished in %.2fs' % (timeElapsed))
    return timeElapsed

# Boyer-Moore


def findMajority(arr, n):
    candidate = -1
    votes = 0

    # Finding majority candidate
    for i in range(n):
        if (votes == 0):
            candidate = arr[i]
            votes = 1
        else:
            if (arr[i] == candidate):
                votes += 1
            else:
                votes -= 1
    count = 0

    # Checking if majority candidate occurs more than n/2
    # times
    for i in range(n):
        if (arr[i] == candidate):
            count += 1

    if (count > n // 2):
        return candidate
    else:
        return -1


def loadDataset(name):
    # load the dataset
    return pd.read_csv(name)


def saveDataset(name, data):
    # save the dataset
    return pd.DataFrame(data).to_csv(name)


cli.add_command(train)
cli.add_command(validate)
cli.add_command(vote)

if __name__ == '__main__':
    cli()

File: ml-models/MaxVote/js/MaxVote.js
const tf = require("@tensorflow/tfjs");
require("@tensorflow/tfjs-node");
const yargs = require("yargs");
const { performance } = require("perf_hooks");
const fs = require("fs");

// Vote Command
yargs.command({
  command: "vote",
  builder: {
    inputs: {
      demandOption: true,
      type: "number",
    },
  },

  // Function for train command
  async handler(argv) {
    var startTime = performance.now();

    const inputs = argv.inputs;
    let data = [];

    // Load datasets
    for (let x = 1; x <= inputs; ++x) {
      let dataset = await loadDataset("predict." + x.toString() + ".csv", "0");

      const res = dataset.mapAsync(({ xs, ys }) => {
        return Object.values(ys)[0];
      });
      data.push(await res.toArray());
    }

    // Save CSV
    let csvContent = "";
    csvContent += ",0" + "\r\n";

    // Max Vote
    let result = [];
    for (let x = 0; x < data[0].length; ++x) {
      let vectorRow = [];
      data.forEach((arr) => {
        vectorRow.push(arr[x]);
      });

      let res = findMajority(vectorRow);
      if (res == -1) {
        res = vectorRow[0];
      }
      result.push(res);

      let row = `${x},${res}`;
      csvContent += row + "\r\n";
    }

    const fileName = "predict.results.csv";
    await saveCSVFile(csvContent, fileName);

    var endTime = performance.now();
    var timeElapsed = endTime - startTime;
    console.log(`Voting finished in ${timeElapsed / 1000} s`);
  },
});

// Train
yargs.command({
  command: "train",
  builder: {},

  // Function for train command
  async handler() {
    var startTime = performance.now();

    const fileName = "latest.model.joblib";
    await saveCSVFile("", fileName);

    var endTime = performance.now();
    var timeElapsed = endTime - startTime;
    console.log(`Max Vote in ${timeElapsed / 1000} s`);
  },
});

// Validate
yargs.command({
  command: "validate",
  builder: {},

  // Function for train command
  async handler() {
    var startTime = performance.now();

    const fileName = "validate.results.csv";
    await saveCSVFile("", fileName);

    var endTime = performance.now();
    var timeElapsed = endTime - startTime;
    console.log(`Max Vote in ${timeElapsed / 1000} s`);
  },
});

yargs.parse(); // To set above changes

// Boyer-Moore Max Vote
function findMajority(nums) {
  var count = 0,
    candidate = -1;

  // Finding majority candidate
  for (var index = 0; index < nums.length; index++) {
    if (count == 0) {
      candidate = nums[index];
      count = 1;
    } else {
      if (nums[index] == candidate) count++;
      else count--;
    }
  }

  count = 0;
  for (var index = 0; index < nums.length; index++) {
    if (nums[index] == candidate) count++;
  }
  if (count > nums.length / 2) return candidate;
  return -1;
}

// Load dataset
async function loadDataset(name, label) {
  //const url = `file://${__dirname}/${name}`;
  const url = `file://./${name}`;
  const dataSet = tf.data.csv(url, {
    hasHeader: true,
    configuredColumnsOnly: true,
    columnConfigs: {
      [label]: {
        required: true,
        isLabel: true,
      },
    },
  });

  return dataSet;
}

// Write File
async function saveCSVFile(data, name) {
  fs.writeFile(name, data, (err) => {
    if (err) return console.error(err);
    console.log("Data saved to disk.");
  });
}
